src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,225 @@
+/*!
+ * Copyright 2019 by XGBoost Contributors
+ */
+#include <thrust/functional.h>
+#include <thrust/random.h>
+#include <thrust/transform.h>
+#include <xgboost/host_device_vector.h>
+#include <xgboost/logging.h>
+
+#include <algorithm>
+
+#include "../../common/compressed_iterator.h"
+#include "../../common/random.h"
+#include "gradient_based_sampler.cuh"
+
+namespace xgboost {
+namespace tree {
+
+GradientBasedSampler::GradientBasedSampler(BatchParam batch_param,
+                                           EllpackInfo info,
+                                           size_t n_rows,
+                                           float subsample,
+                                           SamplingMethod sampling_method)
+    : batch_param_(batch_param), info_(info), sampling_method_(sampling_method) {
+  monitor_.Init("gradient_based_sampler");
+
+  // If `subsample` is not specified, try to figure out how many rows to sample based on available
+  // free GPU memory.
+  if (subsample == 0.0f || subsample == 1.0f) {
+    sample_rows_ = MaxSampleRows(n_rows);
+  } else {
+    sample_rows_ = n_rows * subsample;
+  }
+
+  // If there is enough GPU memory to keep all the rows, don't sample.
+  if (sample_rows_ >= n_rows) {
+    sampling_method_ = kNoSampling;
+    sample_rows_ = n_rows;
+    LOG(CONSOLE) << "Keeping " << sample_rows_ << " rows in GPU memory, not sampling";
+  } else {
+    LOG(CONSOLE) << "Sampling " << sample_rows_ << " rows";
+  }
+
+  // Create a new ELLPACK page with empty rows.
+  page_.reset(new EllpackPageImpl(batch_param.gpu_id, info, sample_rows_));
+
+  // Allocate GPU memory for sampling.
+  if (sampling_method_ != kNoSampling) {
+    ba_.Allocate(batch_param_.gpu_id,
+                 &gpair_, sample_rows_,
+                 &row_weight_, n_rows,
+                 &row_index_, n_rows,
+                 &sample_row_index_, n_rows);
+    thrust::copy(thrust::counting_iterator<size_t>(0),
+                 thrust::counting_iterator<size_t>(n_rows),
+                 dh::tbegin(row_index_));
+  }
+}
+
+// Determine the maximum number of rows that can fit into the available GPU memory.
+size_t GradientBasedSampler::MaxSampleRows(size_t n_rows) {
+  size_t available_memory = dh::AvailableMemory(batch_param_.gpu_id);
+  // Subtract row_weight_, row_index_, and sample_row_index_.
+  available_memory -= n_rows * (sizeof(float) + 2 * sizeof(size_t));
```

Comment:
if `available_memory` is less to accommodate even these indices for the number of rows in the dataset, this can return a negative number which will then be casted into an unsigned value resulting in a weird 2 complement value messing with the estimation.

Reply:
Added a check.